File: app/src/main/java/d/d/meshenger/util/StatsReportUtil.java
package d.d.meshenger.util;

import org.webrtc.RTCStats;
import org.webrtc.RTCStatsReport;

import java.math.BigInteger;
import java.util.Map;

/**
 * @Author Vadym Vikulin
 * @Date 2021/10/10
 * @desc
 */
public class StatsReportUtil {

    private final static int STATS_INTERVAL_MS = 5000;
    private BigInteger lastBytesReceived = BigInteger.valueOf(0);
    private BigInteger lastBytesSent = BigInteger.valueOf(0);
    private Long lastFrameDecoded = Long.valueOf(0);

    public String getStatsReport(RTCStatsReport report) {
        if (report == null) return "";
        String codecId = null;
        String codec = "";
        long bytesSR = 0;
        long sentBytesSR = 0;
        long width = 0, height = 0;
        long frameRate = 0;
        for (RTCStats stats : report.getStatsMap().values()) {
            if (stats.getType().equals("inbound-rtp")) {
                Map<String, Object> members = stats.getMembers();
                if (members.get("mediaType").equals("video")) {
                    codecId = (String) members.get("codecId");
                    BigInteger bytes = (BigInteger) members.get("bytesReceived");
                    bytesSR = bytes.longValue() - lastBytesReceived.longValue();
                    lastBytesReceived = bytes;

                    long currentFrame = (long) members.get("framesDecoded");
                    long lastFrame = lastFrameDecoded;
                    frameRate = (currentFrame - lastFrame) * 1000
                            / STATS_INTERVAL_MS;
                    lastFrameDecoded = currentFrame;
                }
            }
            if (stats.getType().equals("outbound-rtp")) {
                Map<String, Object> members = stats.getMembers();
                if (members.get("mediaType").equals("video")) {
                    //codecId = (String) members.get("codecId");
                    BigInteger bytes = (BigInteger) members.get("bytesSent");
                    sentBytesSR = bytes.longValue() - lastBytesSent.longValue();
                    lastBytesSent = bytes;
                }
            }
            if (stats.getType().equals("track")) {
                Map<String, Object> members = stats.getMembers();
                if (members.get("kind").equals("video")) {
                    width = members.get("frameWidth") == null ? 0 : (long) members.get(
                            "frameWidth");
                    height = members.get("frameHeight") == null ? 0 : (long) members.get(
                            "frameHeight");
                }
            }
        }
        if (codecId != null) {
            codec = (String) report.getStatsMap().get(codecId).getMembers().get("mimeType");
        }

        final String statsReport = "Codec: " + codec
                + "\nResolution: " + width + "x" + height
                + "\nBitrate ↓: " + bytesSR * 8 / STATS_INTERVAL_MS + "kbps"
                + "\nBitrate ↑: " + sentBytesSR * 8 / STATS_INTERVAL_MS + "kbps"
                + "\nFrameRate: " + frameRate;
        return statsReport;
    }
}

File: app/src/main/kotlin/org/rivchain/cuplink/call/StatsReportUtil.kt
package org.rivchain.cuplink.call

import org.webrtc.RTCStatsReport
import java.math.BigInteger

/**
 * @Author Vadym Vikulin
 * @Date 2021/10/10
 * @desc
 */
class StatsReportUtil {
    private var lastBytesReceivedVideo = BigInteger.ZERO
    private var lastBytesSentVideo = BigInteger.ZERO
    private var lastBytesReceivedAudio = BigInteger.ZERO
    private var lastBytesSentAudio = BigInteger.ZERO
    fun getStatsReport(report: RTCStatsReport?): String {
        if (report == null) return ""
        var codecIdVideo: String? = null
        var codecIdAudio: String? = null
        var codecVideo: String? = ""
        var codecAudio: String? = ""
        var receivedBytesSRVideo: Long = 0
        var sentBytesSRVideo: Long = 0
        var receivedBytesSRAudio: Long = 0
        var sentBytesSRAudio: Long = 0
        var widthIn: Long = 0
        var heightIn: Long = 0
        var widthOut: Long = 0
        var heightOut: Long = 0
        var frameRateIn: Long = 0
        var frameRateOut: Long = 0
        for (stats in report.statsMap.values) {
            if (stats.type == "inbound-rtp") {
                val members = stats.members
                if (members["kind"] == "video") {
                    codecIdVideo = members["codecId"] as String?
                    val bytes = members["bytesReceived"] as BigInteger?
                    receivedBytesSRVideo = bytes!!.toLong() - lastBytesReceivedVideo.toLong()
                    lastBytesReceivedVideo = bytes
                    frameRateIn = ((members["framesPerSecond"] as Double?) ?: 0.0).toLong()
                    widthIn = if (members["frameWidth"] == null) 0 else members["frameWidth"] as Long
                    heightIn = if (members["frameHeight"] == null) 0 else members["frameHeight"] as Long
                }
                if (members["kind"] == "audio") {
                    codecIdAudio = members["codecId"] as String?
                    val bytes = members["bytesReceived"] as BigInteger?
                    receivedBytesSRAudio = bytes!!.toLong() - lastBytesReceivedAudio.toLong()
                    lastBytesReceivedAudio = bytes
                }
            }
            if (stats.type == "outbound-rtp") {
                val members = stats.members
                if (members["kind"] == "video") {
                    val bytes = members["bytesSent"] as BigInteger?
                    sentBytesSRVideo = bytes!!.toLong() - lastBytesSentVideo.toLong()
                    lastBytesSentVideo = bytes
                    frameRateOut = ((members["framesPerSecond"] as Double?) ?: 0.0).toLong()
                    widthOut = if (members["frameWidth"] == null) 0 else members["frameWidth"] as Long
                    heightOut = if (members["frameHeight"] == null) 0 else members["frameHeight"] as Long
                }
                if (members["kind"] == "audio") {
                    val bytes = members["bytesSent"] as BigInteger?
                    sentBytesSRAudio = bytes!!.toLong() - lastBytesSentAudio.toLong()
                    lastBytesSentAudio = bytes
                }
            }
        }
        if (codecIdVideo != null) {
            codecVideo = report.statsMap[codecIdVideo]!!.members["mimeType"] as String?
        }
        if (codecIdAudio != null) {
            codecAudio = report.statsMap[codecIdAudio]!!.members["mimeType"] as String?
        }
        return " Codecs\n" +
               "  $codecVideo\n" +
               "  $codecAudio\n" +
               " Quality ↓ ${widthIn}x$heightIn@$frameRateIn\n" +
               " Quality ↑ ${widthOut}x$heightOut@$frameRateOut\n" +
               " Bitrate ⎚ ↓ ${receivedBytesSRVideo * 8 / STATS_INTERVAL_MS}kbps\n" +
               " Bitrate ⎚ ↑ ${sentBytesSRVideo * 8 / STATS_INTERVAL_MS}kbps\n" +
               " Bitrate 🔊 ↓ ${receivedBytesSRAudio * 8 / STATS_INTERVAL_MS}kbps\n" +
               " Bitrate 🔊 ↑ ${sentBytesSRAudio * 8 / STATS_INTERVAL_MS}kbps"
    }

    companion object {
        const val STATS_INTERVAL_MS = 5000L
    }
}